src/logger/winston.ts

```diff
@@ -0,0 +1,64 @@
+import {AbstractLogger, LogLevel} from "./interface";
+import {createLogger, Logger, transports, format} from 'winston';
+
+class WinstonLogger extends AbstractLogger {
+
+  private winston: Logger;
+
+  public constructor() {
+    super();
+    this.winston = createLogger({
+      level: LogLevel.INFO,
+      transports: [
+        new transports.Console({
+          format: format.combine(
+            format.colorize(),
+            format.timestamp({
+              format: 'YYYY-MM-DD HH:mm:ss'
+            }),
+            format.printf(info => `${info.timestamp} ${info.level}: ${info.message}`)
+          ),
+        }),
+
+      ],
+      exitOnError: false
+    });
+  }
+
+  public debug(message: string | object, context?: object): void {
+    this.createLogEntry(LogLevel.DEBUG, message, context);
+  }
+
+  public error(message: string | object, context?: object): void {
+    this.createLogEntry(LogLevel.ERROR, message, context);
+  }
+
+  public info(message: string | object, context?: object): void {
+    this.createLogEntry(LogLevel.INFO, message, context);
+  }
+
+  public warn(message: string | object, context?: object): void {
+    this.createLogEntry(LogLevel.WARN, message, context);
+  }
+
+  private createLogEntry(level: LogLevel, message: string | object, context: object = {}): void {
+    if (typeof message === 'object') {
+      this.winston.log(level, JSON.stringify(message));
+    } else {
+      this.winston.log(level, message, context);
+    }
+  }
+
+  public setLogLevel(level: LogLevel): void {
+    this.winston.level = level;
+  }
+
+  public silent(silent: boolean): void {
+    this.winston.silent = silent;
+  }
+
+}
+
+const logger = new WinstonLogger();
```

Comment:
Valid point. I had this debate a few times. I'm leaning towards singleton because of it's convenience people tend to log often. Passing logger as prop everywhere tends to get awkward.

I'm not really sure what would you configure in each service. only thing that comes to mind is putting class name in log output but it can be easily solved with:
```javascript
//in service
logger.info('some message', this);

//in logger 
meta.constructor.name and print it to log
``` 
Can you think on some service specific configuration?

Reply:
I was thinking about how we could support user-configurable transports.

I'm not really sure that's really needed now, just pondering it. I'm fine just moving forward using the singleton. And if we _really_ need more configurability, we can address it later.